src/validation.cairo

```diff
@@ -38,11 +39,98 @@ impl BlockValidatorImpl of BlockValidator {
 }
 
 #[generate_trait]
-impl TransactionValidatorImpl of TransactionValidator {
+pub impl TransactionValidatorImpl of TransactionValidator {
+    // marker, flag, and witness fields in segwit transactions are not included
+    // this means txid computation is the same for legacy and segwit tx
     fn txid(self: @Transaction) -> u256 {
-        // TODO: implement
-        0
+        let version: felt252 = (*self.version).into();
+        let inputs_count: felt252 = (*self.inputs).len().into();
+        let inputs: Span<TxIn> = *self.inputs;
+        let outputs_count: felt252 = (*self.outputs).len().into();
+        let outputs: Span<TxOut> = *self.outputs;
+        let locktime: felt252 = (*self.lock_time).into();
+
+        // append version (1 byte)
+        let mut sha256_input: ByteArray = "";
+        sha256_input.append_word(version, 1);
+
+        // append padding (3 bytes)
+        let padding: felt252 = 0;
+        sha256_input.append_word(padding, 3);
+
+        // append inputs count
+        sha256_input.append_word(inputs_count, 1);
+
+        // append inputs
+        let mut i = 0;
+        // != instead of < to avoid range check builtin usage for comparison
+        while i != inputs.len() {
```

Comment:
Better way to iterate over the array/span:
```
while let Option::Some(x) = arr.pop_front() {
  sum += x;
};
```

Reply:
way more elegant, thanks.